File: frontend/app/lib/domain/entities/time_range.dart
import 'package:equatable/equatable.dart';

/// Predefined time range options for dashboard filtering
enum TimeRangePreset {
  thisMonth,
  lastMonth,
  last30Days,
  last90Days,
  custom;

  /// Human-readable display name
  String get displayName {
    switch (this) {
      case TimeRangePreset.thisMonth:
        return 'This Month';
      case TimeRangePreset.lastMonth:
        return 'Last Month';
      case TimeRangePreset.last30Days:
        return 'Last 30 Days';
      case TimeRangePreset.last90Days:
        return 'Last 90 Days';
      case TimeRangePreset.custom:
        return 'Custom';
    }
  }

  /// Backend API value
  String get apiValue {
    switch (this) {
      case TimeRangePreset.thisMonth:
        return 'THIS_MONTH';
      case TimeRangePreset.lastMonth:
        return 'LAST_MONTH';
      case TimeRangePreset.last30Days:
        return 'LAST_30_DAYS';
      case TimeRangePreset.last90Days:
        return 'LAST_90_DAYS';
      case TimeRangePreset.custom:
        return 'CUSTOM';
    }
  }
}

/// Represents a date range for filtering metrics
class TimeRange extends Equatable {
  final DateTime start;
  final DateTime end;
  final TimeRangePreset preset;

  const TimeRange({
    required this.start,
    required this.end,
    required this.preset,
  });

  /// Creates a TimeRange for "This Month" (start of month to today)
  factory TimeRange.thisMonth() {
    final now = DateTime.now();
    final start = DateTime(now.year, now.month, 1);
    return TimeRange(
      start: start,
      end: now,
      preset: TimeRangePreset.thisMonth,
    );
  }

  /// Creates a TimeRange for "Last Month" (full previous month)
  factory TimeRange.lastMonth() {
    final now = DateTime.now();
    final firstOfThisMonth = DateTime(now.year, now.month, 1);
    final lastMonth = firstOfThisMonth.subtract(const Duration(days: 1));
    final start = DateTime(lastMonth.year, lastMonth.month, 1);
    return TimeRange(
      start: start,
      end: lastMonth,
      preset: TimeRangePreset.lastMonth,
    );
  }

  /// Creates a TimeRange for "Last 30 Days"
  factory TimeRange.last30Days() {
    final now = DateTime.now();
    final start = now.subtract(const Duration(days: 29));
    return TimeRange(
      start: start,
      end: now,
      preset: TimeRangePreset.last30Days,
    );
  }

  /// Creates a TimeRange for "Last 90 Days"
  factory TimeRange.last90Days() {
    final now = DateTime.now();
    final start = now.subtract(const Duration(days: 89));
    return TimeRange(
      start: start,
      end: now,
      preset: TimeRangePreset.last90Days,
    );
  }

  /// Creates a custom TimeRange
  factory TimeRange.custom(DateTime start, DateTime end) {
    return TimeRange(
      start: start,
      end: end,
      preset: TimeRangePreset.custom,
    );
  }

  /// Creates a TimeRange from a preset
  factory TimeRange.fromPreset(TimeRangePreset preset) {
    switch (preset) {
      case TimeRangePreset.thisMonth:
        return TimeRange.thisMonth();
      case TimeRangePreset.lastMonth:
        return TimeRange.lastMonth();
      case TimeRangePreset.last30Days:
        return TimeRange.last30Days();
      case TimeRangePreset.last90Days:
        return TimeRange.last90Days();
      case TimeRangePreset.custom:
        // Default to last 30 days for custom
        return TimeRange.last30Days();
    }
  }

  /// Number of days in this range (inclusive)
  int get days => end.difference(start).inDays + 1;

  /// Format start date as YYYY-MM-DD
  String get startFormatted =>
      '${start.year}-${start.month.toString().padLeft(2, '0')}-${start.day.toString().padLeft(2, '0')}';

  /// Format end date as YYYY-MM-DD
  String get endFormatted =>
      '${end.year}-${end.month.toString().padLeft(2, '0')}-${end.day.toString().padLeft(2, '0')}';

  /// Human-readable date range
  String get displayRange {
    final startStr =
        '${_monthAbbrev(start.month)} ${start.day}, ${start.year}';
    final endStr = '${_monthAbbrev(end.month)} ${end.day}, ${end.year}';
    return '$startStr - $endStr';
  }

  String _monthAbbrev(int month) {
    const months = [
      'Jan',
      'Feb',
      'Mar',
      'Apr',
      'May',
      'Jun',
      'Jul',
      'Aug',
      'Sep',
      'Oct',
      'Nov',
      'Dec'
    ];
    return months[month - 1];
  }

  @override
  List<Object?> get props => [start, end, preset];
}

File: backend/internal/domain/valueobject/time_range.go
package valueobject

import (
	"time"
)

// TimeRangePreset represents predefined time range options
type TimeRangePreset string

const (
	TimeRangeThisMonth  TimeRangePreset = "THIS_MONTH"
	TimeRangeLastMonth  TimeRangePreset = "LAST_MONTH"
	TimeRangeLast30Days TimeRangePreset = "LAST_30_DAYS"
	TimeRangeLast90Days TimeRangePreset = "LAST_90_DAYS"
	TimeRangeCustom     TimeRangePreset = "CUSTOM"
)

func (t TimeRangePreset) String() string {
	return string(t)
}

func (t TimeRangePreset) IsValid() bool {
	switch t {
	case TimeRangeThisMonth, TimeRangeLastMonth, TimeRangeLast30Days, TimeRangeLast90Days, TimeRangeCustom:
		return true
	}
	return false
}

// DisplayName returns a human-readable name for the preset
func (t TimeRangePreset) DisplayName() string {
	switch t {
	case TimeRangeThisMonth:
		return "This Month"
	case TimeRangeLastMonth:
		return "Last Month"
	case TimeRangeLast30Days:
		return "Last 30 Days"
	case TimeRangeLast90Days:
		return "Last 90 Days"
	case TimeRangeCustom:
		return "Custom"
	default:
		return string(t)
	}
}

// DateRange represents a start and end date range
type DateRange struct {
	Start time.Time
	End   time.Time
}

// NewDateRange creates a new DateRange with validation
func NewDateRange(start, end time.Time) DateRange {
	// Normalize to start of day in UTC
	start = truncateToDay(start)
	end = truncateToDay(end)

	// Swap if start > end
	if start.After(end) {
		start, end = end, start
	}

	return DateRange{Start: start, End: end}
}

// Days returns the number of days in the range (inclusive)
func (d DateRange) Days() int {
	return int(d.End.Sub(d.Start).Hours()/24) + 1
}

// PreviousPeriod returns a DateRange of equal length ending just before this range starts
func (d DateRange) PreviousPeriod() DateRange {
	days := d.Days()
	prevEnd := d.Start.AddDate(0, 0, -1)
	prevStart := prevEnd.AddDate(0, 0, -(days - 1))
	return DateRange{Start: prevStart, End: prevEnd}
}

// Contains checks if a date falls within this range (inclusive)
func (d DateRange) Contains(date time.Time) bool {
	date = truncateToDay(date)
	return !date.Before(d.Start) && !date.After(d.End)
}

// DateRangeForPreset calculates the date range for a given preset
func DateRangeForPreset(preset TimeRangePreset, now time.Time) DateRange {
	now = truncateToDay(now)

	switch preset {
	case TimeRangeThisMonth:
		start := time.Date(now.Year(), now.Month(), 1, 0, 0, 0, 0, time.UTC)
		return DateRange{Start: start, End: now}

	case TimeRangeLastMonth:
		firstOfThisMonth := time.Date(now.Year(), now.Month(), 1, 0, 0, 0, 0, time.UTC)
		end := firstOfThisMonth.AddDate(0, 0, -1)
		start := time.Date(end.Year(), end.Month(), 1, 0, 0, 0, 0, time.UTC)
		return DateRange{Start: start, End: end}

	case TimeRangeLast30Days:
		start := now.AddDate(0, 0, -29) // 30 days including today
		return DateRange{Start: start, End: now}

	case TimeRangeLast90Days:
		start := now.AddDate(0, 0, -89) // 90 days including today
		return DateRange{Start: start, End: now}

	case TimeRangeCustom:
		// For custom, return last 30 days as default
		start := now.AddDate(0, 0, -29)
		return DateRange{Start: start, End: now}

	default:
		// Default to this month
		start := time.Date(now.Year(), now.Month(), 1, 0, 0, 0, 0, time.UTC)
		return DateRange{Start: start, End: now}
	}
}

// truncateToDay normalizes a time to the start of day in UTC
func truncateToDay(t time.Time) time.Time {
	return time.Date(t.Year(), t.Month(), t.Day(), 0, 0, 0, 0, time.UTC)
}
